viewer/packages/metrics/src/FpsLogger.ts

```diff
@@ -0,0 +1,84 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+import { MetricsLogger } from './MetricsLogger';
+export class FpsLogger {
+  private _wasPreviousAnimationFrameRendered = false;
+  private _renderedFramesPerCameraMovement = 0;
+  private _renderedFramesStartTime = 0;
+  private _averageFpsPerMovementSum = 0;
+  private _movementsCount = 0;
+  private _canvasVisibility = true;
+
+  private _isDisposed = false;
+
+  constructor() {
+    document.addEventListener('visibilitychange', this.onRevealSessionEnd);
+  }
+
+  updateCanvasVisibility(isCanvasVisible: boolean): void {
+    if (this._canvasVisibility && !isCanvasVisible && !this._isDisposed) {
+      this.trackAverageSessionFps();
+    }
+
+    this._canvasVisibility = isCanvasVisible;
+  }
+
+  tickCurrentAnimationFrame(isRendered: boolean): void {
+    if (!isRendered) {
+      const movementFps = this.calculateAverageFpsForCurrentMovement();
+
+      if (movementFps) {
+        this._averageFpsPerMovementSum += movementFps;
+        this._movementsCount++;
+      }
+
+      this._renderedFramesPerCameraMovement = 0;
+      this._wasPreviousAnimationFrameRendered = false;
+    }
+
+    if (this._wasPreviousAnimationFrameRendered) {
+      this._renderedFramesPerCameraMovement++;
+    } else {
+      this._renderedFramesStartTime = Date.now();
+      this._wasPreviousAnimationFrameRendered = true;
+    }
```

Comment:
I think I get the gist of the logic here - I'd just like it to be factored into more functions, to make it easier to see what happens :+1:

Reply:
Refactored